broker-daemon/block-order-worker/index.js

```diff
@@ -326,6 +327,53 @@ class BlockOrderWorker extends EventEmitter {
     }
   }
 
+  /**
+   * Applies listeners to a created OrderStateMachine
+   * @private
+   * @param  {OrderStateMachine} osm        State machine to apply listeners to
+   * @param  {BlockOrder} blockOrder Block Order associated with the state machine
+   * @return {OrderStateMachine}
+   */
+  applyOsmListeners (osm, blockOrder) {
+    // Try to complete the entire block order once an underlying order completes
+    osm.once('complete', async () => {
+      try {
+        await this.checkBlockOrderCompletion(blockOrder.id)
+      } catch (e) {
+        this.logger.error(`BlockOrder failed to be completed from order`, { id: blockOrder.id, error: e.stack })
+      }
+      osm.removeAllListeners()
+    })
+
+    // if the fill for this order isn't for the entire order, re-place the remainder
+    osm.once('execute', async () => {
+      this.logger.debug(`Order ${osm.order.orderId} has been filled, re-placing the remainder`)
+      try {
+        const remainingBaseAmount = Big(osm.order.baseAmount).minus(osm.order.fillAmount)
+        // if there is no remaining base amount (i.e. the entire order was filled)
+        // take no action as the block order completion is handled in the `on('complete')` listener
+        if (remainingBaseAmount.gt(0)) {
+          this.logger.debug(`Re-placing an order for ${remainingBaseAmount.toString()} for Block Order ${blockOrder.id}`)
+          await this.workBlockOrder(blockOrder, remainingBaseAmount)
```

Comment:
@treygriffith Im a little confused. Shouldn't `workBlockOrder` be handling this? Or are we saying that `workBlockOrder` will set up the initial orders, but if we get to execute and something has failed, we will need to rerun `workBlockOrder`?

Reply:
Before this change, `workBlockOrder` (or more specifically, `workLimitBlockOrder` and `workMarketBlockOrder`) determined the depth to fill by looking at the total depth of the block order.

With this change, those methods now take a parameter to specify the limit so that we can do things like re-place orders with less the initial amount of the block order.

The other alternative is to fetch all the orders associated with this block order from the store and make a holistic determination about how much is left on the order. That seemed inefficient and unnecessary to me, which is why I did it this way.

Reply:
The was I understand it is that `workBlockOrder` will try to fill as much as possible and then just return. Why would we not add the placement of an order in `workBlockOrder` instead of on the `execute` action?

Reply:
To take our offline discussion online: you are thinking of when initially working the order making sure that you first fill any orders at the given price before placing the remainder. That already happens (in `workLimitBlockOrder`), this is handling the scenario where a counterparty fills part, but not all, of your own placed order.